src/canvas/Canvas.ts

```diff
@@ -1460,193 +1458,126 @@ export class Canvas extends SelectableCanvas {
       !!target &&
       target.selectable &&
-      // if all pre-requisite pass, the target is either something different from the current
-      // activeObject or if an activeSelection already exists
-      // TODO at time of writing why `activeObject.type === 'activeSelection'` matter is unclear.
-      // is a very old condition uncertain if still valid.
-      (activeObject !== target || activeObject.type === 'activeSelection') &&
-      //  make sure `activeObject` and `target` aren't ancestors of each other
-      !target.isDescendantOf(activeObject) &&
-      !activeObject.isDescendantOf(target) &&
+      // group target and active object only if they are different objects
+      // else we try to find a subtarget of `ActiveSelection`
+      (activeObject !== target || isAS) &&
+      //  make sure `activeObject` and `target` aren't ancestors of each other in case `activeObject` is not `ActiveSelection`
+      // if it is then we want to remove `target` from it
+      (isAS ||
+        (!target.isDescendantOf(activeObject) &&
+          !activeObject.isDescendantOf(target))) &&
       //  target accepts selection
-      !target.onSelect({ e: e })
-    );
-  }
-
-  /**
-   * Handles active selection creation for user event
-   * @private
-   * @param {TPointerEvent} e Event object
-   * @param {FabricObject} target
-   */
-  _handleGrouping(e: TPointerEvent, target: FabricObject) {
-    let groupingTarget: FabricObject | undefined = target;
-    // Called always a shouldGroup, meaning that we can trust this._activeObject exists.
-    const activeObject = this._activeObject!;
-    // avoid multi select when shift click on a corner
-    if (activeObject.__corner) {
-      return;
-    }
-    if (groupingTarget === activeObject) {
-      // if it's a group, find target again, using activeGroup objects
-      groupingTarget = this.findTarget(e, true);
-      // if even object is not found or we are on activeObjectCorner, bail out
-      if (!groupingTarget || !groupingTarget.selectable) {
-        return;
-      }
-    }
-    if (activeObject && activeObject.type === 'activeSelection') {
-      this._updateActiveSelection(e, groupingTarget);
-    } else {
-      this._createActiveSelection(e, groupingTarget);
-    }
-  }
-
-  /**
-   * @private
-   */
-  _updateActiveSelection(e: TPointerEvent, target: FabricObject) {
-    const activeSelection = this._activeObject! as ActiveSelection,
-      currentActiveObjects = activeSelection.getObjects();
-    if (target.group === activeSelection) {
-      activeSelection.remove(target);
-      this._hoveredTarget = target;
-      this._hoveredTargets = this.targets.concat();
-      if (activeSelection.size() === 1) {
-        // activate last remaining object
-        this._setActiveObject(activeSelection.item(0) as FabricObject, e);
+      !target.onSelect({ e }) &&
+      // make sure we are not on top of a control
+      !activeObject.__corner
+    ) {
```

Comment:
squashed `_shouldGroup` because of the side effect and weird condition for sub target check